### PR TITLE
chore: Upgrade datafusion -- template

### PR DESCRIPTION
**Pull request by alamb**
Draft as it is waiting for
- [x] https://github.com/apache/arrow-datafusion/pull/4885
- [x] https://github.com/apache/arrow-datafusion/issues/4943

# Note
I don't intend to merge this particular PR as is as it has too many datafusion changes at once and I would like to avoid a giant change all at once.

Instead I am using this PR to debug the issues with DataFusion and then I will plan to split this PR into several piees and upgrades incrementally after backporting various parts of the PR

# Rationale
We are several seeks behind on DataFusion upgrades https://github.com/influxdata/influxdb_iox/issues/6581

# Changes:
1. Update to https://github.com/apache/arrow-datafusion/pull/4952

### CODE REVIEW

File: query_tests/cases/in/dedup_and_predicates_parquet_ingester.expected [hunk 1]
```diff
@@ -8,24 +8,24 @@
...
+|               |     ProjectionExec: expr=[bar@0 as bar, foo@1 as foo, tag@2 as tag, time@3 as time]                                                                                                                |
+|               |       RepartitionExec: partitioning=RoundRobinBatch(4)                                                                                                                                             |
+|               |         DeduplicateExec: [tag@2 ASC,time@3 ASC]                                                                                                                                                    |
+|               |           SortPreservingMergeExec: [tag@2 ASC,time@3 ASC]                                                                                                                                          |
+|               |             SortExec: [tag@2 ASC,time@3 ASC]                                                                                                                                                       |
```
Comment (alamb) [inline query_tests/cases/in/dedup_and_predicates_parquet_ingester.expected:23]: this sort is too high -- is resorting the parquet files

File: query_tests/cases/in/dedup_and_predicates_parquet_ingester.expected [hunk 2]
```diff
@@ -41,15 +41,14 @@
...
+|               |       DeduplicateExec: [tag@2 ASC,time@3 ASC]                                                                                                                                                                                                                                                       |
+|               |         SortPreservingMergeExec: [tag@2 ASC,time@3 ASC]                                                                                                                                                                                                                                             |
+|               |           SortExec: [tag@2 ASC,time@3 ASC]                                                                                                                                                                                                                                                          |
```
Comment (alamb) [inline query_tests/cases/in/dedup_and_predicates_parquet_ingester.expected:48]: sort was moved up incorrectly -- it is also too high

**Review by alamb — COMMENTED** [review]
These plans are almost good -- but there are some sorts that have been moved up unecessairly.

File: query_tests/cases/in/dedup_and_predicates_parquet.expected [hunk 3]
```diff
@@ -33,23 +33,22 @@
...
-|               |   CoalesceBatchesExec: target_batch_size=4096                                                                                                                                                                                                                                                       |
-|               |     FilterExec: tag@2 = A                                                                                                                                                                                                                                                                           |
-|               |       RepartitionExec: partitioning=RoundRobinBatch(4)                                                                                                                                                                                                                                              |
```
Comment (alamb) [inline query_tests/cases/in/dedup_and_predicates_parquet.expected:45]: it is not great to have lost a repartition exec here as the filter could in theory have been done with more parallelism, but the filtering is pushed down all the way to the parquet_exec so I expect the amount of work needed for the final filter is relatively minimal

**Comment by crepererum** [comment]
DF upgrade is done now.

**Pull request closed without merging**